utPLSQL-cli run causes java.sql.SQLException: Locale not recognized

Good day!

I downloaded latest distribution of utPLSQL-cli and tried to execute it on my local environment. I copied ojdbc8.jar as was written in README and even orai18n.jar from Oracle website to utPLSQL-cli/bin folder. However I am receiving the following error:

`$ ./utplsql run tech_user/1@ora12c`
java.sql.SQLException: Locale not recognized
        at oracle.jdbc.driver.T4CTTIoauthenticate.setSessionFields(T4CTTIoauthenticate.java:1493)
        at oracle.jdbc.driver.T4CTTIoauthenticate.<init>(T4CTTIoauthenticate.java:291)
        at oracle.jdbc.driver.T4CConnection.logon(T4CConnection.java:599)
        at oracle.jdbc.driver.PhysicalConnection.connect(PhysicalConnection.java:688)
        at oracle.jdbc.driver.T4CDriverExtension.getConnection(T4CDriverExtension.java:39)
        at oracle.jdbc.driver.OracleDriver.connect(OracleDriver.java:691)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:117)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:375)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:204)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:445)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:516)
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:116)
        at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:97)
        at org.utplsql.cli.ConnectionInfo.getConnection(ConnectionInfo.java:34)
        at org.utplsql.cli.RunCommand.run(RunCommand.java:149)
        at org.utplsql.cli.Cli.main(Cli.java:33)

My NLS_LANG has the following value:

`$ echo $NLS_LANG`
AMERICAN_AMERICA.AL32UTF8

`$ java -version`
java version "1.8.0_77"
Java(TM) SE Runtime Environment (build 1.8.0_77-b03)
Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)

My database parameters:

`select * from v$version;`

--
Oracle Database 12c   Enterprise Edition Release 12.1.0.2.0 - 64bit Production
PL/SQL Release 12.1.0.2.0 -   Production
CORE	12.1.0.2.0	Production
TNS for Linux: Version   12.1.0.2.0 - Production
NLSRTL Version 12.1.0.2.0 -   Production

	

`select * from nls_database_parameters ;`

PARAMETER | VALUE
-- | --
NLS_RDBMS_VERSION | 12.1.0.2.0
NLS_NCHAR_CONV_EXCP | FALSE
NLS_LENGTH_SEMANTICS | BYTE
NLS_COMP | BINARY
NLS_DUAL_CURRENCY | $
NLS_TIMESTAMP_TZ_FORMAT | DD-MON-RR HH.MI.SSXFF AM TZR
NLS_TIME_TZ_FORMAT | HH.MI.SSXFF AM TZR
NLS_TIMESTAMP_FORMAT | DD-MON-RR HH.MI.SSXFF AM
NLS_TIME_FORMAT | HH.MI.SSXFF AM
NLS_SORT | BINARY
NLS_DATE_LANGUAGE | AMERICAN
NLS_DATE_FORMAT | DD-MON-RR
NLS_CALENDAR | GREGORIAN
NLS_NUMERIC_CHARACTERS | .,
NLS_NCHAR_CHARACTERSET | AL16UTF16
NLS_CHARACTERSET | AL32UTF8
NLS_ISO_CURRENCY | AMERICA
NLS_CURRENCY | $
NLS_TERRITORY | AMERICA
NLS_LANGUAGE | AMERICAN

I would greatly appreciate any help or suggestions.

Thank you in advance.

Hi, @jgebal 

Frankly speaking, that was my first choice, and I spent quite a while on it, but didn't manage to make it work, so I gave up. =)

If you have some time I would appreciate if you take a look at the project: https://github.com/nikita-mospan/liquibase-deploy-demo where I was playing around with liquibase and your plugin.

Liquibase is working fine, however when I run 

`mvn test`

[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building liquibase-deploy-demo 1.0
[INFO] ------------------------------------------------------------------------
Downloading: https://maven.oracle.com/org/utplsql/utplsql-maven-plugin/3.1.0-SNAPSHOT/maven-metadata.xml
[INFO]
[INFO] --- liquibase-maven-plugin:3.2.0:update (liquibase-update) @ liquibase-deploy-demo ---
[INFO] ------------------------------------------------------------------------
[INFO] there are no resolved artifacts for the Maven project.
[INFO] Executing on Database: jdbc:oracle:thin:@ora12c:1521/orcl
INFO 12.09.18 14:25: liquibase: Successfully acquired change log lock
INFO 12.09.18 14:25: liquibase: Reading from DATABASECHANGELOG
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_table.sql::nikita.mospan: SQL in file src/main/database/latest/synonym/create-tech_log_table.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_table.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::create-tech_log_table.sql::nikita.mospan ran successfully in 25ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_table_seq.sql::nikita.mospan: SQL in file src/main/database/latest/synonym/create-tech_log_table_seq.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_table_seq.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::create-tech_log_table_seq.sql::nikita.mospan ran successfully in 5ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_instances.sql::nikita.mospan: SQL in file src/main/database/latest/synonym/create-tech_log_instances.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-tech_log_instances.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::create-tech_log_instances.sql::nikita.mospan ran successfully in 3ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-log_table_to_public.sql::nikita.mospan: SQL in file src/main/database/latest/grant/grant-log_table_to_public.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-log_table_to_public.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::grant-log_table_to_public.sql::nikita.mospan ran successfully in 29ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-log_instances_to_public.sql::nikita.mospan: SQL in file src/main/database/latest/grant/grant-log_instances_to_public.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-log_instances_to_public.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::grant-log_instances_to_public.sql::nikita.mospan ran successfully in 6ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-seq_log_table_to_public.sql::nikita.mospan: SQL in file src/main/database/latest/grant/grant-seq_log_table_to_public.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-seq_log_table_to_public.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::grant-seq_log_table_to_public.sql::nikita.mospan ran successfully in 5ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::pk_util_log.pks::nikita.mospan: SQL in file src/main/database/latest/package/pk_util_log.pks executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::pk_util_log.pks::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::pk_util_log.pks::nikita.mospan ran successfully in 13ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::pk_util_log.pkb::nikita.mospan: SQL in file src/main/database/latest/package/pk_util_log.pkb executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::pk_util_log.pkb::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::pk_util_log.pkb::nikita.mospan ran successfully in 13ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-pk_util_log.sql::nikita.mospan: SQL in file src/main/database/latest/synonym/create-pk_util_log.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::create-pk_util_log.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::create-pk_util_log.sql::nikita.mospan ran successfully in 4ms
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-pk_util_log_to_public.sql::nikita.mospan: SQL in file src/main/database/latest/grant/grant-pk_util_log_to_public.sql executed
INFO 12.09.18 14:25: liquibase: src/main/database/changelog/master.xml: src/main/database/changelog/latest/changelog.xml::grant-pk_util_log_to_public.sql::nikita.mospan: ChangeSet src/main/database/changelog/latest/changelog.xml::grant-pk_util_log_to_public.sql::nikita.mospan ran successfully in 5ms
INFO 12.09.18 14:25: liquibase: Successfully released change log lock
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO]
[INFO] --- utplsql-maven-plugin:3.1.0-SNAPSHOT:test (default) @ liquibase-deploy-demo ---
[ERROR]
**java.sql.SQLException: No suitable driver found for jdbc:oracle:thin:@ora12c:1521:orcl**
        at java.sql.DriverManager.getConnection(DriverManager.java:689)
        at java.sql.DriverManager.getConnection(DriverManager.java:247)
        at org.utplsql.maven.plugin.UtPLSQLMojo.execute(UtPLSQLMojo.java:143)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 23.646 s
[INFO] Finished at: 2018-09-12T14:25:20+03:00
[INFO] Final Memory: 18M/196M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.utplsql:utplsql-maven-plugin:3.1.0-SNAPSHOT:test (default) on project liquibase-deploy-demo: No suitable driver found for jdbc:oracle:thin:@ora12c:1521:orcl -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

I even explicitly added dependency com.oracle.ojdbc7 to utplsql-maven-plugin but with no luck.

May be I am missing some configuration options.

I will be glad to discuss this on your project's page, but probably it is not relevant for utPLSQL-cli.

Thank you.

--- a/src/main/java/org/utplsql/cli/LocaleInitializer.java
+++ b/src/main/java/org/utplsql/cli/LocaleInitializer.java
@@ -1,5 +1,7 @@
 package org.utplsql.cli;
 
+import org.utplsql.api.EnvironmentVariableUtil;
+
 import java.util.Locale;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -7,10 +9,9 @@ import java.util.regex.Pattern;
 /** This class makes sure the java locale is set according to the environment variables LC_ALL and LANG
  * We experienced that, in some cases, the locale was not set as expected, therefore this class implements some clear
  * rules:
- *   1. If environment variable NLS_LANG is set, we try to parse its content and set locale according to its value if valid
- *   2. If environment variable LC_ALL is set, we try to parse its content and set locale according to its value if valid
- *   3. If environment variable LANG is set, we try to parse its content and set locale according to its value if valid
- *   4. Otherwise we use default locale
+ *   1. If environment variable LC_ALL is set, we try to parse its content and set locale according to its value if valid
+ *   2. If environment variable LANG is set, we try to parse its content and set locale according to its value if valid
+ *   3. Otherwise we use default locale
  *
  *   @author pesse
  */
@@ -23,12 +24,9 @@ class LocaleInitializer {
      */
     static void initLocale() {
 
-        boolean localeChanged = setDefaultLocale(System.getenv("NLS_LANG"));
-
+        boolean localeChanged = setDefaultLocale(EnvironmentVariableUtil.getEnvValue("LC_ALL"));
         if ( !localeChanged )
-            localeChanged = setDefaultLocale(System.getenv("LC_ALL"));
-        if ( !localeChanged )
-            setDefaultLocale(System.getenv("LANG"));
+            setDefaultLocale(EnvironmentVariableUtil.getEnvValue("LANG"));
     }
 
     /** Set the default locale from a given string like LC_ALL or LANG environment variable

--- a/src/main/java/org/utplsql/cli/datasource/TestedDataSourceProvider.java
+++ b/src/main/java/org/utplsql/cli/datasource/TestedDataSourceProvider.java
@@ -1,6 +1,7 @@
 package org.utplsql.cli.datasource;
 
 import com.zaxxer.hikari.HikariDataSource;
+import org.utplsql.api.EnvironmentVariableUtil;
 import org.utplsql.cli.ConnectionConfig;
 import org.utplsql.cli.exception.DatabaseConnectionFailed;
 
@@ -9,6 +10,8 @@ import java.sql.Connection;
 import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public class TestedDataSourceProvider {
 
@@ -32,12 +35,13 @@ public class TestedDataSourceProvider {
 
         HikariDataSource ds = new HikariDataSource();
 
+        setInitSql(ds);
         testAndSetJdbcUrl(ds);
 
         return ds;
     }
 
-    public void testAndSetJdbcUrl( HikariDataSource ds ) throws SQLException
+    private void testAndSetJdbcUrl( HikariDataSource ds ) throws SQLException
     {
         List<String> errors = new ArrayList<>();
         Throwable lastException = null;
@@ -53,6 +57,33 @@ public class TestedDataSourceProvider {
 
         errors.forEach(System.out::println);
         throw new DatabaseConnectionFailed(lastException);
+    }
+
+    private void setInitSql( HikariDataSource ds ) {
+        String nls_lang = EnvironmentVariableUtil.getEnvValue("NLS_LANG");
+
+        if ( nls_lang != null ) {
+            Pattern pattern = Pattern.compile("^([a-zA-Z ]+)?_?([a-zA-Z ]+)?\\.?([a-zA-Z0-9]+)?$");
+            Matcher matcher = pattern.matcher(nls_lang);
+
+            List<String> sqlCommands = new ArrayList<>(2);
+            if (matcher.matches()) {
+                if ( matcher.group(1) != null)
+                    sqlCommands.add(String.format("ALTER SESSION SET NLS_LANGUAGE='%s'", matcher.group(1)));
+                if ( matcher.group(2) != null)
+                    sqlCommands.add(String.format("ALTER SESSION SET NLS_TERRITORY='%s'", matcher.group(2)));
+
+                if ( sqlCommands.size() > 0 ) {
+                    StringBuilder sb = new StringBuilder();
+                    sb.append("BEGIN\n");
+                    for (String command : sqlCommands)
+                        sb.append(String.format("EXECUTE IMMEDIATE q'[%s]';\n", command));
+                    sb.append("END;");
+
+                    ds.setConnectionInitSql(sb.toString());
+                }
+            }
+        }
     }
 
     private static class ThickConnectStringPossibility implements ConnectStringPossibility {

--- a/src/main/java/org/utplsql/cli/datasource/TestedDataSourceProvider.java
+++ b/src/main/java/org/utplsql/cli/datasource/TestedDataSourceProvider.java
@@ -5,7 +5,6 @@ import org.utplsql.api.EnvironmentVariableUtil;
 import org.utplsql.cli.ConnectionConfig;
 import org.utplsql.cli.exception.DatabaseConnectionFailed;
 
-import java.io.File;
 import java.sql.Connection;
 import java.sql.SQLException;
 import java.util.ArrayList;
@@ -19,7 +18,6 @@ public class TestedDataSourceProvider {
         String getConnectString(ConnectionConfig config);
         String getMaskedConnectString(ConnectionConfig config);
     }
-
 
     private final ConnectionConfig config;
     private List<ConnectStringPossibility> possibilities = new ArrayList<>();
@@ -35,13 +33,13 @@ public class TestedDataSourceProvider {
 
         HikariDataSource ds = new HikariDataSource();
 
-        setInitSql(ds);
-        testAndSetJdbcUrl(ds);
+        setInitSqlFrom_NLS_LANG(ds);
+        setThickOrThinJdbcUrl(ds);
 
         return ds;
     }
 
-    private void testAndSetJdbcUrl( HikariDataSource ds ) throws SQLException
+    private void setThickOrThinJdbcUrl(HikariDataSource ds ) throws SQLException
     {
         List<String> errors = new ArrayList<>();
         Throwable lastException = null;
@@ -59,7 +57,7 @@ public class TestedDataSourceProvider {
         throw new DatabaseConnectionFailed(lastException);
     }
 
-    private void setInitSql( HikariDataSource ds ) {
+    private void setInitSqlFrom_NLS_LANG(HikariDataSource ds ) {
         String nls_lang = EnvironmentVariableUtil.getEnvValue("NLS_LANG");
 
         if ( nls_lang != null ) {

--- a/src/test/java/org/utplsql/cli/DataSourceProviderIT.java
+++ b/src/test/java/org/utplsql/cli/DataSourceProviderIT.java
@@ -5,8 +5,12 @@ import org.utplsql.cli.datasource.TestedDataSourceProvider;
 
 import javax.sql.DataSource;
 import java.io.IOException;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
 import java.sql.SQLException;
 
+import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
 public class DataSourceProviderIT {
@@ -19,5 +23,45 @@ public class DataSourceProviderIT {
         DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
 
         assertNotNull(dataSource);
+    }
+
+    @Test
+    public void initNlsLang() throws SQLException {
+        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
+        System.setProperty("NLS_LANG", "BRAZILIAN PORTUGUESE_BRAZIL.WE8ISO8859P1");
+
+
+        DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
+
+        assertNotNull(dataSource);
+
+        try ( Connection con = dataSource.getConnection() ) {
+            try (PreparedStatement stmt = con.prepareStatement("select value from nls_session_parameters where parameter = 'NLS_LANGUAGE'")) {
+                ResultSet rs = stmt.executeQuery();
+                if ( rs.next() ) {
+                    assertEquals("BRAZILIAN PORTUGUESE", rs.getString(1));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void initPartialNlsLang() throws SQLException {
+        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
+        System.setProperty("NLS_LANG", "_SOMALIA");
+
+
+        DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
+
+        assertNotNull(dataSource);
+
+        try ( Connection con = dataSource.getConnection() ) {
+            try (PreparedStatement stmt = con.prepareStatement("select value from nls_session_parameters where parameter = 'NLS_TERRITORY'")) {
+                ResultSet rs = stmt.executeQuery();
+                if ( rs.next() ) {
+                    assertEquals("SOMALIA", rs.getString(1));
+                }
+            }
+        }
     }
 }

--- a/src/test/java/org/utplsql/cli/DataSourceProviderIT.java
+++ b/src/test/java/org/utplsql/cli/DataSourceProviderIT.java
@@ -12,55 +12,67 @@ import java.sql.SQLException;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.fail;
 
-public class DataSourceProviderIT {
+class DataSourceProviderIT {
 
     @Test
-    public void connectToDatabase() throws IOException, SQLException {
-
-        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
-
-        DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
+    void connectToDatabase() throws SQLException {
+        DataSource dataSource = getDataSource();
 
         assertNotNull(dataSource);
     }
 
     @Test
-    public void initNlsLang() throws SQLException {
-        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
+    void initNlsLang() throws SQLException {
         System.setProperty("NLS_LANG", "BRAZILIAN PORTUGUESE_BRAZIL.WE8ISO8859P1");
-
-
-        DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
+        DataSource dataSource = getDataSource();
 
         assertNotNull(dataSource);
-
-        try ( Connection con = dataSource.getConnection() ) {
-            try (PreparedStatement stmt = con.prepareStatement("select value from nls_session_parameters where parameter = 'NLS_LANGUAGE'")) {
-                ResultSet rs = stmt.executeQuery();
-                if ( rs.next() ) {
-                    assertEquals("BRAZILIAN PORTUGUESE", rs.getString(1));
-                }
-            }
-        }
+        checkNlsSessionParameter(dataSource, "NLS_LANGUAGE", "BRAZILIAN PORTUGUESE");
+        checkNlsSessionParameter(dataSource, "NLS_TERRITORY", "BRAZIL");
     }
 
     @Test
-    public void initPartialNlsLang() throws SQLException {
-        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
+    void initPartialNlsLangTerritory() throws SQLException {
         System.setProperty("NLS_LANG", "_SOMALIA");
-
-
-        DataSource dataSource = new TestedDataSourceProvider(config).getDataSource();
+        DataSource dataSource = getDataSource();
 
         assertNotNull(dataSource);
+        checkNlsSessionParameter(dataSource, "NLS_TERRITORY", "SOMALIA");
+    }
 
+    @Test
+    void initPartialNlsLangLanguage() throws SQLException {
+        System.setProperty("NLS_LANG", "HINDI");
+        DataSource dataSource = getDataSource();
+
+        assertNotNull(dataSource);
+        checkNlsSessionParameter(dataSource, "NLS_LANGUAGE", "HINDI");
+    }
+
+    @Test
+    void initNlsLangEmpty() throws SQLException {
+        System.setProperty("NLS_LANG", "");
+        DataSource dataSource = getDataSource();
+
+        assertNotNull(dataSource);
+    }
+
+    private DataSource getDataSource() throws SQLException {
+        ConnectionConfig config = new ConnectionConfig(TestHelper.getConnectionString());
+        return new TestedDataSourceProvider(config).getDataSource();
+    }
+
+    private void checkNlsSessionParameter( DataSource dataSource, String parameterName, String expectedValue ) throws SQLException {
         try ( Connection con = dataSource.getConnection() ) {
-            try (PreparedStatement stmt = con.prepareStatement("select value from nls_session_parameters where parameter = 'NLS_TERRITORY'")) {
+            try (PreparedStatement stmt = con.prepareStatement("select value from nls_session_parameters where parameter = ?")) {
+                stmt.setString(1, parameterName);
                 ResultSet rs = stmt.executeQuery();
-                if ( rs.next() ) {
-                    assertEquals("SOMALIA", rs.getString(1));
-                }
+                if ( rs.next() )
+                    assertEquals(expectedValue, rs.getString(1));
+                else
+                    fail("Could not get NLS Session parameter value for '" + parameterName + "'");
             }
         }
     }

